sklearn/cluster/tests/test_k_means.py

```diff
@@ -771,4 +774,85 @@ def test_x_squared_norms_init_centroids():
 def test_max_iter_error():
 
     km = KMeans(max_iter=-1)
-    assert_raise_message(ValueError, 'Number of iterations should be', km.fit, X)
+    assert_raise_message(ValueError, 'Number of iterations should be',
+                         km.fit, X)
+
+
+def test_kmeans_float32_64():
+    km = KMeans(n_init=1, random_state=11)
+
+    inertia = {}
+    X_new = {}
+
+    for dtype in [np.float64, np.float32]:
+        X_test = dtype(X)
+        km.fit(X_test)
+        # dtype of cluster centers has to be the dtype of the input data
+        assert_equal(km.cluster_centers_.dtype, dtype)
+        inertia[dtype] = km.inertia_
+        X_new[dtype] = km.transform(km.cluster_centers_)
+        # make sure predictions correspond to the correct label
+        assert_equal(km.predict(X_test[0]), km.labels_[0])
+
+    # compare arrays with low precision since the difference between
+    # 32 and 64 bit sometimes makes a difference up to the 4th decimal place
+    assert_array_almost_equal(inertia[np.float32], inertia[np.float64],
+                              decimal=4)
+    assert_array_almost_equal(X_new[np.float32], X_new[np.float64], decimal=4)
+
+    for dtype in [np.float32, np.float64]:
+        X_csr_test = sp.csr_matrix(X_csr, dtype=dtype)
+        km.fit(X_csr_test)
+        # dtype of cluster centers has to be the dtype of the input data
+        assert_equal(km.cluster_centers_.dtype, dtype)
+        inertia[dtype] = km.inertia_
+        X_new[dtype] = km.transform(km.cluster_centers_)
+        # make sure predictions correspond to the correct label
+        assert_equal(km.predict(X_csr_test[0]),
+                                km.labels_[0])
+
+    assert_array_almost_equal(inertia[np.float32], inertia[np.float64],
+                              decimal=4)
+    assert_array_almost_equal(X_new[np.float32], X_new[np.float64])
+
+
+def test_mb_k_means_float32_64():
```

Comment:
It looks like this is identical to the one above, except for the estimator. Again, we should probably use a loop (or a [generator](http://nose.readthedocs.io/en/latest/writing_tests.html#test-generators))